flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java

```diff
@@ -183,6 +235,42 @@ public void notifyCheckpointComplete(long checkpointId, OperatorChain<?, ?> oper
 		env.getTaskStateManager().notifyCheckpointComplete(checkpointId);
 	}
 
+	@Override
+	public void notifyCheckpointAborted(long checkpointId, OperatorChain<?, ?> operatorChain, Supplier<Boolean> isRunning) throws Exception {
+
+		if (isRunning.get()) {
+			LOG.debug("Notification of aborted checkpoint for task {}", taskName);
+			// only happens when the task always received checkpoints to abort but never trigger or executing.
+			if (abortedCheckpointIds.size() >= maxRecordAbortedCheckpoints) {
+				abortedCheckpointIds.pollFirst();
```

Comment:
Why do we remove the oldest checkpoint? Why not the one that was _canceled first_?
I guess both are valid options, but for the 2nd we could use `LinkedHashMap` with O(1) time complexity.

Reply:
I think this suggestion sounds reasonable, changed to `ArrayDeque`.